plugin/pkg/scheduler/equivalence_cache.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+import (
+	"github.com/golang/groupcache/lru"
+	"hash/adler32"
+
+	"k8s.io/kubernetes/pkg/api"
+	hashutil "k8s.io/kubernetes/pkg/util/hash"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/algorithm"
+	schedulerapi "k8s.io/kubernetes/plugin/pkg/scheduler/api"
+	"sync"
+)
+
+const maxCacheEntries = 4096
+
+type HostPredicate struct {
+	Fit         bool
+	FailReasons []algorithm.PredicateFailureReason
+}
+
+type AlgorithmCache struct {
+	predicatesCache *lru.Cache
+	prioritiesCache *lru.Cache
+}
+
+func newAlgorithmCache() AlgorithmCache {
+	return AlgorithmCache{
+		predicatesCache: lru.New(maxCacheEntries),
+		prioritiesCache: lru.New(maxCacheEntries),
+	}
+}
+
+// Store a map of predicate cache with maxsize
+type EquivalenceCache struct {
+	getEquivalencePod         algorithm.GetEquivalencePodFunc
+	algorithmCache            map[string]AlgorithmCache
+	invalidAlgorithmCacheList sets.String
+	allCacheExpired           bool
+	expireLock                *sync.RWMutex
+}
+
+func NewEquivalenceCache(getEquivalencePodFunc algorithm.GetEquivalencePodFunc) *EquivalenceCache {
+	return &EquivalenceCache{
+		getEquivalencePod:         getEquivalencePodFunc,
+		algorithmCache:            make(map[string]AlgorithmCache),
+		invalidAlgorithmCacheList: sets.NewString(),
+		allCacheExpired:           false,
+		expireLock:                new(sync.RWMutex),
+	}
+}
+
+// addPodPriority adds pod priority for equivalence class
+func (ec *EquivalenceCache) addPodPriority(podKey uint64, nodeName string, score int) {
+	if _, exist := ec.algorithmCache[nodeName]; !exist {
+		ec.algorithmCache[nodeName] = newAlgorithmCache()
+	}
+	ec.algorithmCache[nodeName].prioritiesCache.Add(podKey, schedulerapi.HostPriority{Host: nodeName, Score: score})
+}
```

Comment:
hmm - I actually started wondering about priorities. The problem with priorities is that most of them are global functions (in other words, a priority for a given (pod, node) pair, depends on all the nodes in the system (and what is happening on them). This may be very problematic... @davidopp - thoughts?